### PR TITLE
Upgrade Kubebuilder to 3.0.0 and optimize Github workflow

### PR DESCRIPTION
**Pull request by Jeffwan**
Address #30 #25 https://github.com/ray-project/ray-contrib/issues/24 This change helps resolve CI problem. check #30 for more details.

1. Deprecate genclient and deepcopy-gen markers
2. Add health/ready check and use nonroot user in deployment
3. Fix github workflow issue

Signed-off-by: Jiaxin Shan <seedjeffwan@gmail.com>

/cc @akanso @chenk008 @chaomengyuan

### CODE REVIEW

File: ray-operator/api/v1alpha1/raycluster_types.go [hunk 1]
```diff
@@ -101,10 +101,8 @@ const (
 )
 
 // RayCluster is the Schema for the RayClusters API
-// +genclient
```
Comment (Jeffwan) [inline ray-operator/api/v1alpha1/raycluster_types.go:104]: We can replace the `+k8s:deepcopy-gen:interfaces=...` marker (which is deprecated in kubebuilder) with `+kubebuilder:object:root=true.`

We don’t need the following markers any more (they’re not used anymore, and are relics from much older versions of KubeBuilder):

```
// +genclient
// +k8s:openapi-gen=true
```

source: https://book.kubebuilder.io/migration/legacy/migration_guide_v1tov2.html

File: ray-operator/Makefile [hunk 2]
```diff
@@ -1,6 +1,6 @@
 
 # Image URL to use all building/pushing image targets
-IMG ?=  ray-controller:latest
```
Comment (Jeffwan) [inline ray-operator/Makefile:3]: Make files comes from kubebuilder v3.0.0 and it gives more commands to use

File: .github/workflows/test-job.yaml [hunk 3]
```diff
@@ -16,38 +16,36 @@ jobs:
     runs-on: ubuntu-latest
     steps:
 
-    - name: Set up Go 1.x
+    - name: Set up Go 1.15.x
       uses: actions/setup-go@v2
       with:
-        go-version: ^1.13
+        go-version: '~1.15.1'
```
Comment (Jeffwan) [inline .github/workflows/test-job.yaml:22]: pin golang version

File: .github/workflows/test-job.yaml [hunk 4]
```diff
@@ -16,38 +16,36 @@ jobs:
...
 
     - name: Check out code into the Go module directory
       uses: actions/checkout@v2
       with:
-        ref: master
+        # When checking out the repository that
+        # triggered a workflow, this defaults to the reference or SHA for that event.
+        # Default value should work for both pull_request and merge(push) event.
+        ref: ${{github.event.pull_request.head.sha}}
```
Comment (Jeffwan) [inline .github/workflows/test-job.yaml:30]: In my personal testing, this works for push event as well.

File: ray-operator/main.go [hunk 5]
```diff
@@ -21,31 +23,36 @@ var (
...
-		Scheme:             scheme,
-		MetricsBindAddress: metricsAddr,
-		LeaderElection:     enableLeaderElection,
-		Port:               9443,
+		Scheme:                 scheme,
+		MetricsBindAddress:     metricsAddr,
+		Port:                   9443,
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         enableLeaderElection,
+		LeaderElectionID:       "7054f481.",
```
Comment (chenk008) [inline ray-operator/main.go:55]: How about using `ray-operator-leader` as LeaderElectionID?  I think it is leader configmap name.

Comment (Jeffwan) [reply ray-operator/main.go:55]: @chenk008 Sounds good. I will make the change

Comment (Jeffwan) [reply ray-operator/main.go]: Addressed

**Review by chenk008 — APPROVED** [review]

**Comment by Jeffwan** [comment]
I will merge this since it's blocking CI.  @akanso if you can some comments on the  github workflow change, I will address it in separate PR

**Pull request merged**